### ISSUE TITLE
haproxy reverse proxy not working with API

### ISSUE THREAD

**Issue by gitsf** [theotherp/nzbhydra#406]
First off I am sorry if I shouldn't post this here, but I wasn't sure where else to go for support. 

current setup:
Ubuntu Server 16.04
Use Docker containers for all my services: Haproxy, Sonarr, Hydra, Couchpotato

I currently use Haproxy for SSL termination and proxying, so I dont have to open ports for each service they all run behind haproxy over port 443.

I dont like using paths for my services, they seem to require a lot more setup for many of the programs. So my URLs are sonarr.myurl.com & hydra.myurl.com & couchpotato.myurl.com etc
Haproxy looks at the domain name to know what backend to send them to.
All containers are all on their own private docker network, they can all communicate with each other over this.
Hydra seems to be working great from the public url for searching and sending to sabnzbd container.

The problem seems to be when using the API publicly over the internet.
I can use the API just fine over the docker network from sonarr to hydra. 
I can also open the hydra port and skip haproxy and go straight to IP:PORT and it will work just fine over the internet. 
But if I specify https://hydra.myurl.com in sonarr it fails, sonarr says it times out. Any manual searches also fail with no results.
Hydra logs dont even show any of the connection attempts coming in.
Looking at the sonarr logs I have tried loading the following in chrome and it works.(i have modified the actual url and removed the API sonarr is calling: 
https://hydra.myurl.com/api?t=tvsearch&cat=5030,5040&extended=1&apikey=REMOVED&offset=0&limit=100
So not sure how chrome is requesting it differently vs sonarr

Because the API works using the public IP with port forwarding, I believe the problem is something in haproxy breaking it.

backend part of haproxy.cfg:
backend hydra_back
		option forwardfor
		http-request set-header X-Forwarded-For $proxy_add_x_forwarded_for;
		server hydra hydra:5075 check

in hydra I have the external url set to: https://hydra.myurl.com

Has anyone else had any issues with this or gotten this to work with haproxy?

For right now I have my local sonarr setup using the docker private network, but I would like the API to be accessible to a hosted server I have, so I am needing public access to it.

I have tried getting haproxy to log every request and response but I am having trouble getting that to work. It is only logging that it is starting up and backends have started, I cant get anything else to log.

**Comment by gitsf** [comment]
Any of the URLs pulled from sonarr logs where it attempted to search for a show, if I put that into chrome with the API it works. It only appears when sonarr is requesting it that it fails. However it does work from within sonarr if I use the public IP and port forward direct to hydra. Which I dont want to do I would like to run it through haproxy.

**Comment by theotherp** [comment]
Ok, if I understand you correctly then the problem is that Sonarr is unable to communicate with Hydra via the proxy but it works using your browser?

If I remember correctly somebody on /r/usenet had a similar problem and I think solved it by binding Hydra to 127.0.0.1 instead of 0.0.0.0 (or vice versa). I'm no expert in network stuff and don't know Haproxy, so I don't think I can help you there. Perhaps some of the other active users here has an idea. If not you can try posting on /r/usenet.

Sorry, cannot help you with that one.

**Comment by gitsf** [comment]
Thanks I am looking through /r/usenet for some help.
I have now setup couchpotato to use hydra and it is working just fine with the api. Searches come back and nzbs are snatched and sent to sabnzbd. Hydra also shows the search history and snatched history with no issues. This is going through the public IP and haproxy. So confusing so maybe the problem is with sonarr.
Thanks again for all the hard work on this, I had debated about setting this up for weeks and am happy to finally do it. I think this will be great once I get it working with everything.

**Comment by theotherp** [comment]
You could also ask on /r/sonarr or their support page, a lot of people with experience on there.

**Comment by gitsf** [comment]
I have figured out the problem, it is Mono and the SSL it supports.
Sonarr is supposed to fallback to using curl if the Mono connection fails, but that didnt seem to be happening in my docker container.
My Haproxy config was blocking older SSL types, specifically TLSv1.0
Once I changed my haproxy config to allow TLSv1.0 it started working.

Hopefully this will help others who may have certain SSL encryptions disabled in their reverse proxies.
The reverse proxy must allow TLS v1.0 connections in order for Sonarr to connect via https

More info from sonarr forum: [https://forums.sonarr.tv/t/timeouts-to-nzb-hydra-via-public-ip-and-haproy/12475](https://forums.sonarr.tv/t/timeouts-to-nzb-hydra-via-public-ip-and-haproy/12475)
